### DISCUSSION TITLE
Binary i3d export

### DISCUSSION THREAD

**Discussion by StjerneIdioten** [StjerneIdioten/I3D-Blender-Addon#112] [Ideas & Feature Requests]
**Is your feature request related to a problem? Please describe.**
Export the i3d file as binary.

**Comment by kbrandwijk** [comment]
Be aware that GE does more than just save as binary. It also de-duplicates shapes (by hash?) and some other things. So implementing this 'like GE does it' is more  than just a file format task.

**Reply by StjerneIdioten** [reply]
Well I didn't expect it to simply be a matter of "turn it into binary" ;-) What do you mean exactly by de-duplicating? Like compressing the individual shapes themselves? 

And on the other hand there is still the whole legal thing that needs to be handled in some way.

The current state of things is that Giants has been made aware that it would be helpful if GE itself had a parameter for converting a non-binary i3d into a binary i3d, as a sort of way around all this (for the export side of things at least) and to my knowledge they have noted it down, but who knows when they will actually get around to implement it ;-)

**Reply by kbrandwijk** [reply]
Take two identical cubes, and export them to i3d, there will be two shapes defined (assuming you're not using linked duplicates). Now save the i3d and binary i3d in GE. Upon inspection of that i3d, you will notice that both cubes have magically gotten the same shapeId.

**Reply by StjerneIdioten** [reply]
Ah in that sense. 
So similar to what GE does for materials as well (Which I have gotten "bug reports" about plenty of times)
If you create two cubes with two different materials, but with identical settings on the materials, and export them from Blender. Upon saving in GE all of the identical (but differently named) materials, will be collapsed into one material.

In that case I agree that it is probably some form of hashing, or at least that is how I would implement it. I already do something similar to keep vertices within a shape unique.

**Reply by kbrandwijk** [reply]
Yeah I suspect the material combine is a side effect of the shape combine. It combines the shapes, and because it's left with only one shape, it also only keeps the first material? Not 100% sure of the internals, but I've been plagued by it many times before settling for a proper workflow.

**Reply by StjerneIdioten** [reply]
Hmm, maybe. But if I recall correctly it happens regardless of whether or not they shape the same mesh shape. Might have to check that again.

**Comment by ssnd292** [comment]
I have been working on this and an implementation ready.

Waiting for a Giants dev to push a fix to their i3dconverter.exe in their own toolset, then I can setup a merge request.

**Comment by StjerneIdioten** [comment]
[This release](https://github.com/StjerneIdioten/I3D-Blender-Addon/releases/tag/v4.3.0) adds support for exporting i3d files as binary from Blender. It utilises the i3dConverter.exe that the GE addon also uses. I have added support for just using the same one as the GE addon, in case you are utilising both addons. Or if you are only using the community exporter you get the option of downloading it manually or automatically by providing your Giants Developer Network login (The login is NOT saved)
